File: src/personal_data_warehouse_voice_memos/sync.py
from __future__ import annotations

from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
import threading

from personal_data_warehouse_voice_memos.scanner import (
    VoiceMemoFileCandidate,
    VoiceMemoRecording,
    recording_from_candidate,
    scan_voice_memo_file_candidates,
)
from personal_data_warehouse_voice_memos.state import VoiceMemosUploadState

PARTIAL_LOCAL_DURATION_TOLERANCE_SECONDS = 5.0


@dataclass(frozen=True)
class VoiceMemosUploadSummary:
    recordings_seen: int
    recordings_skipped: int
    recordings_uploaded: int
    metadata_uploaded: int
    bytes_seen: int = 0
    bytes_uploaded: int = 0
    bytes_skipped: int = 0
    recordings_selected: int = 0
    recordings_deferred: int = 0


@dataclass(frozen=True)
class _RecordingSyncResult:
    uploaded: int
    skipped: int
    metadata_uploaded: int
    bytes_uploaded: int
    bytes_skipped: int
    audio_present: bool = False
    metadata_present: bool = False


class VoiceMemosUploadRunner:
    def __init__(
        self,
        *,
        account: str,
        recordings_path: Path | str,
        extensions: tuple[str, ...],
        ingest_client,
        logger,
        now=None,
        limit: int | None = None,
        workers: int = 1,
        mode: str = "full",
        upload_state: VoiceMemosUploadState | None = None,
        min_file_age_seconds: int = 0,
        before_upload_check: Callable[[], str | None] | None = None,
        max_upload_bytes: int | None = None,
    ) -> None:
        # Uploads go through the app's ingest endpoints only.
        if ingest_client is None:
            raise ValueError("ingest_client is required")
        # Recordings larger than what the chosen upload route can deliver (the
        # app's cap, or Cloudflare's 100 MiB edge limit on the public host) are
        # deferred rather than retried forever: a single oversized memo would
        # otherwise 413 on every run and, because the failure re-raises, wedge
        # the whole sync so newer memos never upload. Default to the client's
        # advertised ceiling so the limit tracks the route actually in use.
        if max_upload_bytes is None:
            max_upload_bytes = getattr(ingest_client, "effective_max_upload_bytes", None)
        self._max_upload_bytes = max_upload_bytes if (max_upload_bytes and max_upload_bytes > 0) else None
        self._account = account
        self._recordings_path = Path(recordings_path).expanduser()
        self._extensions = extensions
        self._ingest_client = ingest_client
        self._state_lock = threading.Lock()
        self._logger = logger
        self._now = now or (lambda: datetime.now(tz=UTC))
        self._limit = limit
        self._workers = max(1, workers)
        if mode not in {"full", "incremental"}:
            raise ValueError("mode must be 'full' or 'incremental'")
        self._mode = mode
        self._upload_state = upload_state
        self._min_file_age_seconds = max(0, min_file_age_seconds)
        self._before_upload_check = before_upload_check

    def sync(self) -> VoiceMemosUploadSummary:
        self._logger.info(
            "Scanning Voice Memos in %s for extensions: %s",
            self._recordings_path,
            ", ".join(self._extensions),
        )
        candidates = scan_voice_memo_file_candidates(self._recordings_path, extensions=self._extensions)
        if self._mode == "incremental":
            return self._sync_incremental(candidates)

        if self._limit is not None:
            candidates = candidates[: self._limit]
        recordings_seen = len(candidates)
        bytes_seen = sum(candidate.size_bytes for candidate in candidates)
        partial_candidates = [candidate for candidate in candidates if voice_memo_candidate_is_partially_materialized(candidate)]
        partial_candidate_set = set(partial_candidates)
        candidates = [candidate for candidate in candidates if candidate not in partial_candidate_set]
        for candidate in partial_candidates:
            self._logger.warning(
                "Deferring %s because Voice Memos reports %.2fs total but only %.2fs local audio",
                candidate.filename,
                candidate.duration_seconds or 0,
                candidate.local_duration_seconds or 0,
            )
        recordings = [(candidate, recording_from_candidate(candidate)) for candidate in candidates]
        self._logger.info(
            "Found %s Voice Memos recordings totaling %s",
            recordings_seen,
            format_bytes(bytes_seen),
        )
        if recordings and self._before_upload_check is not None:
            skip_reason = self._before_upload_check()
            if skip_reason:
                self._logger.warning("Skipping Voice Memos upload: %s", skip_reason)
                return VoiceMemosUploadSummary(
                    recordings_seen=recordings_seen,
                    recordings_skipped=0,
                    recordings_uploaded=0,
                    metadata_uploaded=0,
                    bytes_seen=bytes_seen,
                    recordings_selected=len(recordings),
                    recordings_deferred=len(recordings) + len(partial_candidates),
                )
        self._logger.info("Uploading with %s worker(s)", self._workers)

        if self._workers == 1 or len(recordings) <= 1:
            results = [
                self._sync_candidate(index=index, total=len(recordings), candidate=candidate, recording=recording)
                for index, (candidate, recording) in enumerate(recordings, start=1)
            ]
        else:
            with ThreadPoolExecutor(max_workers=self._workers, thread_name_prefix="voice-memos") as executor:
                futures = [
                    executor.submit(
                        self._sync_candidate,
                        index=index,
                        total=len(recordings),
                        candidate=candidate,
                        recording=recording,
                    )
                    for index, (candidate, recording) in enumerate(recordings, start=1)
                ]
                results = [future.result() for future in as_completed(futures)]

        uploaded = sum(result.uploaded for result in results)
        skipped = sum(result.skipped for result in results)
        metadata_uploaded = sum(result.metadata_uploaded for result in results)
        bytes_uploaded = sum(result.bytes_uploaded for result in results)
        bytes_skipped = sum(result.bytes_skipped for result in results)

        summary = VoiceMemosUploadSummary(
            recordings_seen=recordings_seen,
            recordings_skipped=skipped,
            recordings_uploaded=uploaded,
            metadata_uploaded=metadata_uploaded,
            bytes_seen=bytes_seen,
            bytes_uploaded=bytes_uploaded,
            bytes_skipped=bytes_skipped,
            recordings_selected=len(recordings),
            recordings_deferred=len(partial_candidates),
        )
        self._logger.info(
            "Voice Memos upload summary: seen=%s (%s), uploaded=%s (%s), skipped=%s (%s), metadata=%s",
            summary.recordings_seen,
            format_bytes(summary.bytes_seen),
            summary.recordings_uploaded,
            format_bytes(summary.bytes_uploaded),
            summary.recordings_skipped,
            format_bytes(summary.bytes_skipped),
            summary.metadata_uploaded,
        )
        return summary

    def _sync_incremental(self, candidates: list[VoiceMemoFileCandidate]) -> VoiceMemosUploadSummary:
        if self._limit is not None:
            candidates = candidates[: self._limit]
        bytes_seen = sum(candidate.size_bytes for candidate in candidates)
        self._logger.info(
            "Found %s Voice Memos recordings totaling %s",
            len(candidates),
            format_bytes(bytes_seen),
        )
        selected: list[VoiceMemoFileCandidate] = []
        state_skipped = 0
        age_deferred = 0
        partial_deferred = 0
        oversize_deferred = 0
        now = self._now()

        for candidate in candidates:
            if self._min_file_age_seconds and (now - candidate.file_modified_at).total_seconds() < self._min_file_age_seconds:
                age_deferred += 1
                continue
            if self._max_upload_bytes is not None and candidate.size_bytes > self._max_upload_bytes:
                oversize_deferred += 1
                self._logger.warning(
                    "Deferring %s (%s): exceeds the %s upload ceiling for the current route",
                    candidate.filename,
                    format_bytes(candidate.size_bytes),
                    format_bytes(self._max_upload_bytes),
                )
                continue
            if voice_memo_candidate_is_partially_materialized(candidate):
                partial_deferred += 1
                self._logger.warning(
                    "Deferring %s because Voice Memos reports %.2fs total but only %.2fs local audio",
                    candidate.filename,
                    candidate.duration_seconds or 0,
                    candidate.local_duration_seconds or 0,
                )
                continue
            entry = self._upload_state.entry_for(candidate) if self._upload_state is not None else None
            if entry is not None and entry.complete and entry.matches(candidate):
                state_skipped += 1
                continue
            selected.append(candidate)

        self._logger.info(
            "Incremental selection: selected=%s skipped=%s deferred=%s",
            len(selected),
            state_skipped,
            age_deferred + partial_deferred + oversize_deferred,
        )
        if not selected:
            return VoiceMemosUploadSummary(
                recordings_seen=len(candidates),
                recordings_skipped=state_skipped,
                recordings_uploaded=0,
                metadata_uploaded=0,
                bytes_seen=bytes_seen,
                bytes_uploaded=0,
                bytes_skipped=sum(candidate.size_bytes for candidate in candidates if self._is_state_complete(candidate)),
                recordings_selected=0,
                recordings_deferred=age_deferred + partial_deferred + oversize_deferred,
            )

        if self._before_upload_check is not None:
            skip_reason = self._before_upload_check()
            if skip_reason:
                self._logger.warning("Skipping Voice Memos upload: %s", skip_reason)
                return VoiceMemosUploadSummary(
                    recordings_seen=len(candidates),
                    recordings_skipped=state_skipped,
                    recordings_uploaded=0,
                    metadata_uploaded=0,
                    bytes_seen=bytes_seen,
                    bytes_uploaded=0,
                    bytes_skipped=sum(candidate.size_bytes for candidate in candidates if self._is_state_complete(candidate)),
                    recordings_selected=len(selected),
                    recordings_deferred=len(selected) + age_deferred + partial_deferred + oversize_deferred,
                )

        self._logger.info("Uploading with %s worker(s)", self._workers)
        recordings = [(candidate, recording_from_candidate(candidate)) for candidate in selected]
        if self._workers == 1 or len(recordings) <= 1:
            results = [
                self._sync_candidate(index=index, total=len(recordings), candidate=candidate, recording=recording)
                for index, (candidate, recording) in enumerate(recordings, start=1)
            ]
        else:
            with ThreadPoolExecutor(max_workers=self._workers, thread_name_prefix="voice-memos") as executor:
                futures = [
                    executor.submit(
                        self._sync_candidate,
                        index=index,
                        total=len(recordings),
                        candidate=candidate,
                        recording=recording,
                    )
                    for index, (candidate, recording) in enumerate(recordings, start=1)
                ]
                results = [future.result() for future in as_completed(futures)]

        uploaded = sum(result.uploaded for result in results)
        skipped = state_skipped + sum(result.skipped for result in results)
        metadata_uploaded = sum(result.metadata_uploaded for result in results)
        bytes_uploaded = sum(result.bytes_uploaded for result in results)
        bytes_skipped = sum(result.bytes_skipped for result in results) + sum(
            candidate.size_bytes for candidate in candidates if self._is_state_complete(candidate)
        )
        summary = VoiceMemosUploadSummary(
            recordings_seen=len(candidates),
            recordings_skipped=skipped,
            recordings_uploaded=uploaded,
            metadata_uploaded=metadata_uploaded,
            bytes_seen=bytes_seen,
            bytes_uploaded=bytes_uploaded,
            bytes_skipped=bytes_skipped,
            recordings_selected=len(selected),
            recordings_deferred=age_deferred + partial_deferred + oversize_deferred,
        )
        self._logger.info(
            "Voice Memos upload summary: seen=%s (%s), selected=%s, uploaded=%s (%s), skipped=%s (%s), deferred=%s, metadata=%s",
            summary.recordings_seen,
            format_bytes(summary.bytes_seen),
            summary.recordings_selected,
            summary.recordings_uploaded,
            format_bytes(summary.bytes_uploaded),
            summary.recordings_skipped,
            format_bytes(summary.bytes_skipped),
            summary.recordings_deferred,
            summary.metadata_uploaded,
        )
        return summary

    def _sync_candidate(
        self,
        *,
        index: int,
        total: int,
        candidate: VoiceMemoFileCandidate,
        recording: VoiceMemoRecording,
    ) -> _RecordingSyncResult:
        try:
            result = self._sync_recording(index=index, total=total, recording=recording)
            if self._upload_state is not None:
                with self._state_lock:
                    self._upload_state.mark_success(
                        candidate=candidate,
                        content_sha256=recording.content_sha256,
                        audio_uploaded=result.audio_present,
                        metadata_uploaded=result.metadata_present,
                        now=self._now(),
                    )
            return result
        except Exception as exc:
            if self._upload_state is not None:
                with self._state_lock:
                    self._upload_state.mark_failure(
                        candidate=candidate,
                        content_sha256=recording.content_sha256,
                        error=str(exc),
                        now=self._now(),
                    )
            raise

    def _is_state_complete(self, candidate: VoiceMemoFileCandidate) -> bool:
        entry = self._upload_state.entry_for(candidate) if self._upload_state is not None else None
        return bool(entry is not None and entry.complete and entry.matches(candidate))

    def _sync_recording(self, *, index: int, total: int, recording: VoiceMemoRecording) -> _RecordingSyncResult:
        # The app owns object keys, kinds, and pdw_* tags and dedups by content
        # sha, so we always send and let the app collapse duplicates.
        recorded_at = recording.recorded_at.isoformat()
        self._logger.info(
            "[%s/%s] upload %s (%s, sha256=%s) -> app",
            index,
            total,
            recording.filename,
            format_bytes(recording.size_bytes),
            short_sha256(recording.content_sha256),
        )
        self._ingest_client.upload_voice_memo_audio(
            recording.path.read_bytes(),
            recorded_at=recorded_at,
            extension=recording.extension,
            content_type=recording.content_type,
        )
        metadata_payload = build_metadata(
            account=self._account,
            recording=recording,
            uploaded_at=self._now(),
        )
        self._ingest_client.upload_voice_memo_metadata(
            metadata_payload,
            recorded_at=recorded_at,
            audio_content_sha256=recording.content_sha256,
        )
        return _RecordingSyncResult(
            uploaded=1,
            skipped=0,
            metadata_uploaded=1,
            bytes_uploaded=recording.size_bytes,
            bytes_skipped=0,
            audio_present=True,
            metadata_present=True,
        )


def build_metadata(
    *,
    account: str,
    recording: VoiceMemoRecording,
    uploaded_at: datetime,
) -> dict[str, object]:
    recording_payload = {
        "recording_id": recording.recording_id,
        "title": recording.title,
        "original_path": str(recording.path),
        "filename": recording.filename,
        "extension": recording.extension,
        "content_type": recording.content_type,
        "size_bytes": recording.size_bytes,
        "content_sha256": recording.content_sha256,
        "file_created_at": recording.file_created_at.isoformat(),
        "file_modified_at": recording.file_modified_at.isoformat(),
        "recorded_at": recording.recorded_at.isoformat(),
    }
    if recording.duration_seconds is not None:
        recording_payload["duration_seconds"] = recording.duration_seconds
    if recording.local_duration_seconds is not None:
        recording_payload["local_duration_seconds"] = recording.local_duration_seconds
    return {
        "schema_version": 1,
        "source": "apple_voice_memos",
        "account": account,
        "uploaded_at": uploaded_at.isoformat(),
        "recording": recording_payload,
    }


def voice_memo_candidate_is_partially_materialized(candidate: VoiceMemoFileCandidate) -> bool:
    if candidate.duration_seconds is None or candidate.local_duration_seconds is None:
        return False
    if candidate.duration_seconds > 0 and candidate.local_duration_seconds <= 0:
        return True
    return candidate.duration_seconds > candidate.local_duration_seconds + PARTIAL_LOCAL_DURATION_TOLERANCE_SECONDS


def short_sha256(value: str) -> str:
    return value[:12]


def format_bytes(size: int) -> str:
    units = ("B", "KiB", "MiB", "GiB", "TiB")
    value = float(size)
    for unit in units:
        if value < 1024 or unit == units[-1]:
            if unit == "B":
                return f"{int(value)} {unit}"
            return f"{value:.1f} {unit}"
        value /= 1024
    return f"{size} B"

File: app/internal/uploaders/voicememos/sync.go
package voicememos

import (
	"errors"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/ingestclient"
	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

// PartialLocalDurationToleranceSeconds is how far the locally materialized
// audio may fall short of the recording's full duration before the file is
// treated as still downloading from iCloud.
const PartialLocalDurationToleranceSeconds = 5.0

// Summary is the run's counts (VoiceMemosUploadSummary).
type Summary struct {
	RecordingsSeen     int
	RecordingsSkipped  int
	RecordingsUploaded int
	MetadataUploaded   int
	BytesSeen          int64
	BytesUploaded      int64
	BytesSkipped       int64
	RecordingsSelected int
	RecordingsDeferred int
}

// Uploader is what the runner needs from the ingest client.
type Uploader interface {
	UploadVoiceMemoAudio(content []byte, recordedAt, extension, contentType string) (ingestclient.StoredObject, error)
	UploadVoiceMemoMetadata(payload map[string]any, recordedAt, audioContentSHA256 string) (ingestclient.StoredObject, error)
}

// AppKicker makes sure the Voice Memos store is still being fed before the
// scan; the runner calls it first, before trusting what it scanned.
type AppKicker func(recordingsPath string, logger common.Logger) AppKick

// Runner uploads recordings through the app (VoiceMemosUploadRunner).
type Runner struct {
	Account           string
	RecordingsPath    string
	Extensions        []string
	Client            Uploader
	Logger            common.Logger
	Now               func() time.Time
	Limit             int // 0 means no limit
	Workers           int
	Mode              string // "full" or "incremental"
	State             *State
	MinFileAgeSeconds int
	BeforeUploadCheck func() string
	// MaxUploadBytes is the route's ceiling; 0 means unbounded. The CLI sets
	// it from the client's EffectiveMaxUploadBytes.
	MaxUploadBytes int64
	// EnsureApp defaults to EnsureVoiceMemosAppRunning against the process
	// environment; tests inject a no-op.
	EnsureApp AppKicker

	stateMu sync.Mutex
}

type recordingResult struct {
	uploaded         int
	skipped          int
	metadataUploaded int
	bytesUploaded    int64
	bytesSkipped     int64
	audioPresent     bool
	metadataPresent  bool
}

// Sync runs one pass.
func (r *Runner) Sync() (Summary, error) {
	if r.Client == nil {
		return Summary{}, errors.New("ingest_client is required")
	}
	if r.Mode == "" {
		r.Mode = "full"
	}
	if r.Mode != "full" && r.Mode != "incremental" {
		return Summary{}, errors.New("mode must be 'full' or 'incremental'")
	}
	if r.Now == nil {
		r.Now = func() time.Time { return time.Now().UTC() }
	}
	if r.Workers < 1 {
		r.Workers = 1
	}
	if r.MinFileAgeSeconds < 0 {
		r.MinFileAgeSeconds = 0
	}
	if r.Logger == nil {
		r.Logger = &common.RecordingLogger{}
	}
	root := common.ExpandUser(r.RecordingsPath)
	// Before trusting what the store contains, make sure the store is still
	// being fed: with voicememod and the app both quit, macOS stops pulling
	// iCloud changes and every run reads a healthy selected=0.
	kick := r.EnsureApp
	if kick == nil {
		kick = func(path string, logger common.Logger) AppKick {
			return EnsureVoiceMemosAppRunning(path, logger, os.Getenv, "", nil)
		}
	}
	kick(root, r.Logger)
	r.Logger.Infof("Scanning Voice Memos in %s for extensions: %s", root, strings.Join(r.Extensions, ", "))
	candidates, err := ScanCandidates(root, r.Extensions)
	if err != nil {
		return Summary{}, err
	}
	if r.Mode == "incremental" {
		return r.syncIncremental(candidates)
	}
	return r.syncFull(candidates)
}

func (r *Runner) syncFull(candidates []Candidate) (Summary, error) {
	if r.Limit > 0 && len(candidates) > r.Limit {
		candidates = candidates[:r.Limit]
	}
	seen := len(candidates)
	var bytesSeen int64
	for _, c := range candidates {
		bytesSeen += c.SizeBytes
	}
	var partial, kept []Candidate
	for _, c := range candidates {
		if IsPartiallyMaterialized(c) {
			partial = append(partial, c)
		} else {
			kept = append(kept, c)
		}
	}
	for _, c := range partial {
		r.Logger.Warningf("Deferring %s because Voice Memos reports %.2fs total but only %.2fs local audio", c.Filename, floatOrZero(c.Duration), floatOrZero(c.LocalDuration))
	}
	recordings := make([]Recording, 0, len(kept))
	for _, c := range kept {
		recording, err := RecordingFromCandidate(c)
		if err != nil {
			return Summary{}, err
		}
		recordings = append(recordings, recording)
	}
	r.Logger.Infof("Found %d Voice Memos recordings totaling %s", seen, common.FormatBytes(bytesSeen))
	if len(recordings) > 0 && r.BeforeUploadCheck != nil {
		if reason := r.BeforeUploadCheck(); reason != "" {
			r.Logger.Warningf("Skipping Voice Memos upload: %s", reason)
			return Summary{
				RecordingsSeen: seen, BytesSeen: bytesSeen,
				RecordingsSelected: len(recordings), RecordingsDeferred: len(recordings) + len(partial),
			}, nil
		}
	}
	r.Logger.Infof("Uploading with %d worker(s)", r.Workers)
	results, errs := r.uploadAll(recordings)
	summary := Summary{RecordingsSeen: seen, BytesSeen: bytesSeen, RecordingsSelected: len(recordings), RecordingsDeferred: len(partial)}
	for _, err := range errs {
		if err != nil {
			return summary, err
		}
	}
	for _, result := range results {
		summary.RecordingsUploaded += result.uploaded
		summary.RecordingsSkipped += result.skipped
		summary.MetadataUploaded += result.metadataUploaded
		summary.BytesUploaded += result.bytesUploaded
		summary.BytesSkipped += result.bytesSkipped
	}
	r.Logger.Infof("Voice Memos upload summary: seen=%d (%s), uploaded=%d (%s), skipped=%d (%s), metadata=%d",
		summary.RecordingsSeen, common.FormatBytes(summary.BytesSeen), summary.RecordingsUploaded, common.FormatBytes(summary.BytesUploaded),
		summary.RecordingsSkipped, common.FormatBytes(summary.BytesSkipped), summary.MetadataUploaded)
	return summary, nil
}

func (r *Runner) syncIncremental(candidates []Candidate) (Summary, error) {
	if r.Limit > 0 && len(candidates) > r.Limit {
		candidates = candidates[:r.Limit]
	}
	var bytesSeen int64
	for _, c := range candidates {
		bytesSeen += c.SizeBytes
	}
	r.Logger.Infof("Found %d Voice Memos recordings totaling %s", len(candidates), common.FormatBytes(bytesSeen))
	var selected []Candidate
	stateSkipped, ageDeferred, partialDeferred, oversizeDeferred := 0, 0, 0, 0
	now := r.Now()
	for _, c := range candidates {
		if r.MinFileAgeSeconds > 0 && now.Sub(c.FileModifiedAt).Seconds() < float64(r.MinFileAgeSeconds) {
			ageDeferred++
			continue
		}
		if r.MaxUploadBytes > 0 && c.SizeBytes > r.MaxUploadBytes {
			oversizeDeferred++
			r.Logger.Warningf("Deferring %s (%s): exceeds the %s upload ceiling for the current route", c.Filename, common.FormatBytes(c.SizeBytes), common.FormatBytes(r.MaxUploadBytes))
			continue
		}
		if IsPartiallyMaterialized(c) {
			partialDeferred++
			r.Logger.Warningf("Deferring %s because Voice Memos reports %.2fs total but only %.2fs local audio", c.Filename, floatOrZero(c.Duration), floatOrZero(c.LocalDuration))
			continue
		}
		if r.isStateComplete(c) {
			stateSkipped++
			continue
		}
		selected = append(selected, c)
	}
	deferred := ageDeferred + partialDeferred + oversizeDeferred
	r.Logger.Infof("Incremental selection: selected=%d skipped=%d deferred=%d", len(selected), stateSkipped, deferred)
	completeBytes := func() int64 {
		var total int64
		for _, c := range candidates {
			if r.isStateComplete(c) {
				total += c.SizeBytes
			}
		}
		return total
	}
	if len(selected) == 0 {
		return Summary{
			RecordingsSeen: len(candidates), RecordingsSkipped: stateSkipped, BytesSeen: bytesSeen,
			BytesSkipped: completeBytes(), RecordingsDeferred: deferred,
		}, nil
	}
	if r.BeforeUploadCheck != nil {
		if reason := r.BeforeUploadCheck(); reason != "" {
			r.Logger.Warningf("Skipping Voice Memos upload: %s", reason)
			return Summary{
				RecordingsSeen: len(candidates), RecordingsSkipped: stateSkipped, BytesSeen: bytesSeen,
				BytesSkipped: completeBytes(), RecordingsSelected: len(selected), RecordingsDeferred: len(selected) + deferred,
			}, nil
		}
	}
	r.Logger.Infof("Uploading with %d worker(s)", r.Workers)
	recordings := make([]Recording, 0, len(selected))
	for _, c := range selected {
		recording, err := RecordingFromCandidate(c)
		if err != nil {
			return Summary{}, err
		}
		recordings = append(recordings, recording)
	}
	// Per-file failures are collected, not returned mid-batch: one memo that
	// errors must not abort the rest. Successful uploads are recorded in the
	// state, so the next run resumes past them; the first error is returned
	// once at the end so the run still exits non-zero.
	results, errs := r.uploadAll(recordings)
	var failures []error
	summary := Summary{RecordingsSeen: len(candidates), RecordingsSkipped: stateSkipped, BytesSeen: bytesSeen, RecordingsSelected: len(selected), RecordingsDeferred: deferred}
	for i, err := range errs {
		if err != nil {
			r.Logger.Warningf("Failed to upload %s: %s", recordings[i].Filename, err)
			failures = append(failures, err)
			continue
		}
		summary.RecordingsUploaded += results[i].uploaded
		summary.RecordingsSkipped += results[i].skipped
		summary.MetadataUploaded += results[i].metadataUploaded
		summary.BytesUploaded += results[i].bytesUploaded
		summary.BytesSkipped += results[i].bytesSkipped
	}
	summary.BytesSkipped += completeBytes()
	r.Logger.Infof("Voice Memos upload summary: seen=%d (%s), selected=%d, uploaded=%d (%s), skipped=%d (%s), deferred=%d, metadata=%d",
		summary.RecordingsSeen, common.FormatBytes(summary.BytesSeen), summary.RecordingsSelected, summary.RecordingsUploaded,
		common.FormatBytes(summary.BytesUploaded), summary.RecordingsSkipped, common.FormatBytes(summary.BytesSkipped),
		summary.RecordingsDeferred, summary.MetadataUploaded)
	if len(failures) > 0 {
		r.Logger.Warningf("Voice Memos upload finished with %d failed recording(s) after uploading %d; re-raising the first so the run is marked failed (successful uploads are recorded, so the next run resumes past them)", len(failures), summary.RecordingsUploaded)
		return summary, failures[0]
	}
	return summary, nil
}

// uploadAll runs syncCandidate over the recordings with Workers goroutines,
// returning per-recording results and errors in input order.
func (r *Runner) uploadAll(recordings []Recording) ([]recordingResult, []error) {
	results := make([]recordingResult, len(recordings))
	errs := make([]error, len(recordings))
	workers := r.Workers
	if len(recordings) <= 1 {
		workers = 1
	}
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, workers)
	for i, recording := range recordings {
		wg.Add(1)
		semaphore <- struct{}{}
		go func(i int, recording Recording) {
			defer wg.Done()
			defer func() { <-semaphore }()
			results[i], errs[i] = r.syncCandidate(i+1, len(recordings), recording)
		}(i, recording)
	}
	wg.Wait()
	return results, errs
}

func (r *Runner) isStateComplete(c Candidate) bool {
	if r.State == nil {
		return false
	}
	entry, ok := r.State.EntryFor(c)
	return ok && entry.Complete() && entry.Matches(c)
}

func (r *Runner) syncCandidate(index, total int, recording Recording) (recordingResult, error) {
	result, err := r.syncRecording(index, total, recording)
	if r.State != nil {
		r.stateMu.Lock()
		defer r.stateMu.Unlock()
		if err != nil {
			r.State.MarkFailure(recording.Candidate, recording.ContentSHA256, err.Error(), r.Now())
		} else {
			r.State.MarkSuccess(recording.Candidate, recording.ContentSHA256, result.audioPresent, result.metadataPresent, r.Now())
		}
	}
	return result, err
}

func (r *Runner) syncRecording(index, total int, recording Recording) (recordingResult, error) {
	// The app owns object keys, kinds, and pdw_* tags and dedups by content
	// sha, so we always send and let the app collapse duplicates.
	recordedAt := common.ISOFormat(recording.RecordedAt)
	r.Logger.Infof("[%d/%d] upload %s (%s, sha256=%s) -> app", index, total, recording.Filename, common.FormatBytes(recording.SizeBytes), common.ShortSHA256(recording.ContentSHA256))
	content, err := os.ReadFile(recording.Path)
	if err != nil {
		return recordingResult{}, err
	}
	if _, err := r.Client.UploadVoiceMemoAudio(content, recordedAt, recording.Extension, recording.ContentType); err != nil {
		return recordingResult{}, err
	}
	payload := BuildMetadata(r.Account, recording, r.Now())
	if _, err := r.Client.UploadVoiceMemoMetadata(payload, recordedAt, recording.ContentSHA256); err != nil {
		return recordingResult{}, err
	}
	return recordingResult{uploaded: 1, metadataUploaded: 1, bytesUploaded: recording.SizeBytes, audioPresent: true, metadataPresent: true}, nil
}

// BuildMetadata is the recording's JSON sidecar, byte-identical (through
// common.CanonicalJSON) to the Python build_metadata.
func BuildMetadata(account string, recording Recording, uploadedAt time.Time) map[string]any {
	payload := map[string]any{
		"recording_id":     recording.RecordingID,
		"title":            recording.Title,
		"original_path":    recording.Path,
		"filename":         recording.Filename,
		"extension":        recording.Extension,
		"content_type":     recording.ContentType,
		"size_bytes":       recording.SizeBytes,
		"content_sha256":   recording.ContentSHA256,
		"file_created_at":  common.ISOFormat(recording.FileCreatedAt),
		"file_modified_at": common.ISOFormat(recording.FileModifiedAt),
		"recorded_at":      common.ISOFormat(recording.RecordedAt),
	}
	if recording.Duration != nil {
		payload["duration_seconds"] = *recording.Duration
	}
	if recording.LocalDuration != nil {
		payload["local_duration_seconds"] = *recording.LocalDuration
	}
	return map[string]any{
		"schema_version": int64(1),
		"source":         "apple_voice_memos",
		"account":        account,
		"uploaded_at":    common.ISOFormat(uploadedAt),
		"recording":      payload,
	}
}

// IsPartiallyMaterialized reports whether Voice Memos says the recording is
// longer than the audio it has downloaded locally, i.e. iCloud is still
// delivering it and the file on disk is incomplete.
func IsPartiallyMaterialized(c Candidate) bool {
	if c.Duration == nil || c.LocalDuration == nil {
		return false
	}
	if *c.Duration > 0 && *c.LocalDuration <= 0 {
		return true
	}
	return *c.Duration > *c.LocalDuration+PartialLocalDurationToleranceSeconds
}

func floatOrZero(value *float64) float64 {
	if value == nil {
		return 0
	}
	return *value
}
